contracts/utils/math/Math.sol

```diff
@@ -149,4 +149,62 @@ library Math {
         }
         return result;
     }
+
+    /**
+     * @dev Returns the square root of a number.
+     *
+     * Inspired by Henry S. Warren, Jr.'s "Hacker's Delight" (Chapter 11).
+     */
+    function sqrt(uint256 a) internal pure returns (uint256) {
+        if (a == 0) {
+            return 0;
+        }
+
+        // For our first guess, we use the log2 of the square root. We do that by shifting `a` and only counting half
+        // of the bits. the partial result produced is a power of two that verifies `result <= sqrt(a) < 2 * result`
+        uint256 result = 1;
+        uint256 x = a;
+        if (x > 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF) {
```

Comment:
then what about doing `x >= 1 << 128`. Would that also get optimized? to the same bytecode. Because there is no `gte` opcode I'm worried it would be compiled to `not(lt(..., ...))`

Reply:
I think I found an even better rewording.